### DISCUSSION TITLE
What is the correct way of mapping with Star Schema?

### DISCUSSION THREAD

**Discussion by dajemet379** [mapstruct/mapstruct#2898] [Q&A]
I have a use case of mapping where I need to query tables from other DBs to enrich the data of some fields. e.g. to map the field `Integer carSerialNumber` to `String car`, I need to query a Star Schema that holds the correlation between `4721847391` and `Honda Bla 2019` (which is the target value).

I wonder whether that's considered not to be part of the mapping, or that IO operations are a common and regular use case in MapStruct.

**Comment by Zegveld** [comment]
You need to supply the knowledge of how to retrieve it to the mappers that you generate, there are several options for this. Below I've worked out a few of them.

When using a dependency injection framework you can inject a class with the required knowledge.
For example:
```
// implementation of the interface below has the knowledge of how to retrieve the value from the db.
public interface DbResource {
    @org.mapstruct.Named("fromSerial")
    String fromSerialNumber(Integer serial);
}
@Mapper( componentModel = "spring", uses = DbResource.class )
public interface RegistrationMapper {
    @Mapping( source="carSerial", target="carDetails", qualifiedByName="fromSerial" )
    Registration map(RegistrationSerials serials);
}
```
Without a dependency injection framework you can supply it using a Context, but this requires an additional method as the methods from the context aren't picked up for mapping actions:
```
@Mapper
public interface RegistrationMapper {
    @Mapping( source="carSerial", target="carDetails", qualifiedByName="fromSerial" )
    Registration map(RegistrationSerials serials, @Context DbResource resource);

    @Named( "fromSerial" )
    default String fromSerialNumber(Integer serial, @Context DbResource resource) {
        return resource.fromSerialNumber( serial );
    }
}
```
or using an expression:
```
@Mapper
public interface RegistrationMapper {
    @Mapping( target="carDetails", expression="java(resource.fromSerialNumber(serials.getCarSerial()))")
    Registration map(RegistrationSerials serials, @Context DbResource resource);
```

As you can see I'm using the `@Named` annotation from the `org.mapstruct` package. This is to make certain that this method is not used for other mapping actions.
